### ISSUE TITLE
System Compatibility: TDE5 / DSA5

### ISSUE THREAD

**Issue by m1lklinG** [Saibot393/LocknKey#64]
Hi there,

i really like the idea of your module! Unfortunately it seems not to support TDE/DSA at this very moment (https://foundryvtt.com/packages/dsa5). 

Is it possible to make this possible?

Kind regards

**Comment by Saibot393** [comment]
Sure, could you give me quick run down of/a link to the relevant rules?

**Comment by m1lklinG** [comment]
QuickStart Guide can be found here: 
https://paizo.com/products/btpy9maz/discuss?The-Dark-Eye-QuickStart-Rules

To provide a little more information:

Each skill check in TDE/DSA depends on 3 attribute checks, which differ from skill to skill. 
Lockpicking (i.e.) depends on Intuition / Agility / Agility.

Let's say your character has 9 Intuition and 11 Agility and your Skill in Lockpicking is 4, then the check would occur as follows:

1. Roll 3D20 simultaneously (1st D20 against Intuition,  2nd D20 against Agility, 3rd D20 against Agility).
2. Every D20 is meant to be <= the corresponding attribute value. If this is the case in all 3D20 then skill check succeeded.
3. Your skill level is a wildcard / buffer to cope with rolls worse than <= attributeValue. 
4. With the character above you had 4 buffer points (in total, for the whole check!) which can be used to lower the D20 rolls. 
5. If the buffer points are not enough to achieve @Nr. 2, then the skill check failed. 

Last but not least:

Skill checks with two (or three) 1's are critical success checks.
Skill checks with two (or three) 20's are critial failure checks.

**Comment by Saibot393** [comment]
Is there some kind of difficulty scaling for check, e.g. could a lock do something like add to the buffer to make the check easier?

**Comment by m1lklinG** [comment]
Yes, indeed. There is at least one factor which can modify a skill check.
In our case (lockpicking) it could be the difficulty of the lock itself.

This would modify the buffer value for this very lock (LockpickingSkill - LockDifficulty = TempBufferValue). 
The higher the buffer value the better the playing actor is at lockpicking. And vice versa.
There is no auto-success or auto-fail by high or low buffervalues itself, as critical success rolls (at least 2D20 = "1") and critical failure rolls (at least 2D20 = "20") within the skillcheck will overrule. 

> Example (from above) for positive (or "zero") TempBufferValues:
> 
> Characters values: 9 IN, 11 AG, 4 LockPicking
> LockDifficulty 4
> 
> -> Skillcheck: 3D20 (9 / 11 / 11) + BufferValue (4-4=0) => 3D20(9 / 11 / 11)
>
> You'd have to roll <= 9 / <= 11 / <= 11 to pick the lock.
>
>But: SkillCheck = 20 / 20 / X, 20 / X / 20, X / 20 / 20 results in critical failure.

The BufferValue can also become negative.
This has a little different mechanic, as it makes every attribute checkRoll harder.

> Example (from above) for negative TempBufferValues:
> 
> Characters values: 9 IN, 11 AG, 4 LockPicking
> LockDifficulty 6
> 
> -> Skillcheck: 3D20 (9 / 11 / 11) + BufferValue (4-6=**-2**) => 3D20(9**-2** / 11**-2** / 11**-2**)
>
> You'd have to roll <=7 / <= 9 / <= 9 to pick the lock then.
>
>But: SkillCheck = 1 / 1 / X, 1/ X / 1, X / 1/ 1results in critical success.

**Comment by Saibot393** [comment]
I will deffinetly be able to implement the first case you described. The second case where the BufferValue becomes negative and additional values are applied to the dice rolls themselves will be a bit more tricky, but i will see what i can do.

**Comment by m1lklinG** [comment]
It's a pitty I am not a coder at all. 
As far as i understand you will try to redo some of the work which already has been done forr the game system?

The SkillCheck itself within the system TDE/DSA is (of course) working.
Couldn't you "just" make your module "trigger" this SkillCheck and grab the result after it has been rolled?

**Comment by Saibot393** [comment]
Things like rolls are extremely system dependent. In Pf2e for example it is quite easy to let Lock&Key trigger the pick lock action and use the result on the lock. Pf2e is more of an exception though. 
In most systems rolls as such are not even directly available and the only thing i can normally do is open the roll request window, so i have to code my own roll logic based on core foundry rolls. (The reason i can not simply use the roll request window is, that it is quite hard in most systems to grab the results since it is only put in the chat without any further indentifiers, if i am lucky they at least have their used skill defined). 
That is also the reason Lock & Key requires the systems roll formula. The problem is now, that the lock modifier would have to be present in the roll formula for the second case to work, which would be a bit hard to do with the current way lock & key calculates locks. Maybe i can come up with a solution for that though.

**Comment by Saibot393** [comment]
I have discovered another problem. Unlike most systems, DSA5 uses items for its talents/skill. Items are impossible (as far as i know) to reference in roll formulas without hard coding them.  I will continue searching for a solution but it may take a while.

**Comment by Saibot393** [comment]
Good new, the problem has been resolved, i will now start working on the formulas

**Comment by m1lklinG** [comment]
Very cool :-)
If i can help you any further, please give me a hint.

**Comment by Saibot393** [comment]
So, i think i have a working, albeit ugly, roll formula. The formula will currently not work without a small update (including crits). Before i push the update i would like to test of the formula works, so could you test a few scenarios just to make sure that everything works?
This is the formula: `(max(1d20 - (@actor.system.characteristics.in.value + min(@DC, 0)), 0) +max(1d20 - (@actor.system.characteristics.ff.value + min(@DC, 0)),0) + max(1d20 - (@actor.system.characteristics.ff.value + min(@DC, 0)),0)) - (@skills.Pick_Locks.system.talentValue.value + max(@DC, 0))` for lock picking. 
As mentioned the formula currently doesn't work, but will output into the chat by how much the results overshoots the skills value (including negative DCs). So values smaller or equal to 0 will be a success, large than 0 a failure.
The formula also assumes that the lock picking skill is named "Pick Locks", should you use a language other then english you will have to replace "Pick_Locks" in the formula with the appropiate skill name (replacing " " with "_")

**Comment by m1lklinG** [comment]
Hey there. During christmas time i unfortunately couldnt make it to have a look at this - hope you had a few relaxing days as well.

How can i use this formula in foundry? I tried to use it in an empty macro as well as as entry for your "configure lock&key" -> "lockpick formula". It does not seem to do anything yet. There is a high chance i did it very wrong.. sorry :)

**Comment by Saibot393** [comment]
The formula above should simply work when set as the "Lockpick roll formula". Are there any chat messages or console(F12) messgaes when trying to pick a lock while having this formula set?

**Comment by m1lklinG** [comment]
<img width="404" alt="image" src="https://github.com/Saibot393/LocknKey/assets/148483620/0d41bebc-3a1f-4c25-a57c-747a241ebb78">

**Comment by Saibot393** [comment]
Could you, while having an actor selected enter this into the console and send me the ouput:
```
r = new Roll("1d20 + @actor.system.characteristics.in.value", {actor : canvas.tokens.controlled[0].actor});
await r.evaluate();
r.result;
```

**Comment by m1lklinG** [comment]
> '18 + 14'

I did that a few times. 
The first number (above: 18) seems to be randomly generated between 1 and 20. 
The second number (above: 14) seems to be persistent.

**Comment by Saibot393** [comment]
Ok, that what i expected. 
One questions, which Lock & Key version are you using?
To narrow down the problem could you try these things:
1. Enter this line into the console while controlling a token and send me the output: `canvas.tokens.controlled[0].actor.items.filter(item => ["skill"].includes(item.type))`
2. Could you try using these roll formula for the "Lockpick roll formula" and try using lockpicking on a door:
"1d20 - (@actor.system.characteristics.in.value + min(@DC, 0))"
3. Could you also try this formula:
"@skills.Pick_Locks.system.talentValue.value"

**Comment by m1lklinG** [comment]
I'm using v. 2.4.3

> 1. Enter this line into the console while controlling a token and send me the output: `canvas.tokens.controlled[0].actor.items.filter(item => ["skill"].includes(item.type))`

<img width="391" alt="image" src="https://github.com/Saibot393/LocknKey/assets/148483620/646a6ce3-3cd8-4d71-91e5-fbb68702ad16">

> 2. Could you try using these roll formula for the "Lockpick roll formula" and try using lockpicking on a door:
>    `1d20 - (@actor.system.characteristics.in.value + min(@DC, 0))`

<img width="390" alt="image" src="https://github.com/Saibot393/LocknKey/assets/148483620/a80d4296-e28a-4bbf-81fd-85645e3d3e67">

> 3. Could you also try this formula:
>    `@skills.Pick_Locks.system.talentValue.value`

This one does open (and lock, if already open) the lock! 
But it seems like there is no roll being requested.

**Comment by Saibot393** [comment]
Sorry for all the question but i am really lost on that first error message you send.
What foundry and wat DSA5 version are you using?
Could you send me file of the world in which the first error appeared?

**Comment by m1lklinG** [comment]
Foundry v. 11.315
DSA5 v. 5.2.8

[world.test.zip](https://github.com/Saibot393/LocknKey/files/13823695/world.test.zip)

**Comment by Saibot393** [comment]
Thank you, but i do not seem to be able to load the world file you attached, or rather i don't know how to load it. It only seems to contain the .json and .bak, but lacks the worlds files (scenes, settings, actors...)

**Comment by Saibot393** [comment]
Thank you, but i do not seem to be able to load the world file you attached, or rather i don't know how to load it. It only seems to contain the .json and .bak, but lacks the worlds files (scenes, settings, actors...)

**Comment by m1lklinG** [comment]
Ah - it is a backup file.. i thought you could load the world out of it. 
There you go:

[test.zip](https://github.com/Saibot393/LocknKey/files/13823774/test.zip)

**Comment by Saibot393** [comment]
Yes, i think discord may be more convenient, if you wan't you can DM me: saibot5279

**Comment by Dicio83** [comment]
I have only just found the module and I am thrilled. I also need it for DSA 5. Is there a new news?

**Comment by Saibot393** [comment]
I did a few tests with m1lklinG and i think i should have all the bugs narrowed down. I would estimate that i push and update for DSA5 within the next one or two weeks

**Comment by Dicio83** [comment]
Ok, are spells that allow locks to be opened also built in? Can they be prised open later with crowbars or lock picks?

If you need help trying it out, please write to me. I can't programme, but I can try things out :-)

**Comment by Saibot393** [comment]
Lock picking spells are relatively easy to add (though will need to be configured by hand). You will basically just need to enter a unique string present in the flavour text of these spells so that lock & key can recognise them.
Both lock pciking and using raw strength to break locks will be possible, though i just noticed that i am not sure which kind of skill would be used to break open locks, i would guess "Feat of strength"?

**Comment by Dicio83** [comment]
Thats right - Feath of Strength is correct

**Comment by Saibot393** [comment]
I have released support for DSA with v2.4.3. I will leave this issue open for now, in case something doesn't work as intended or there are questions

**Comment by Dicio83** [comment]
Youre awsome! Will test it and response. Where is your donation-link? :-)

**Comment by Dicio83** [comment]
This appears wenn I try to lock pick
![image](https://github.com/Saibot393/LocknKey/assets/116867654/0ce1391e-ec53-4931-bf70-92da4715bd1c)

and this when I try to use raw strenght

![image](https://github.com/Saibot393/LocknKey/assets/116867654/099ae15f-4983-45f3-be77-85738926e698)

**Comment by Dicio83** [comment]
Ok. Changed "Lock_pick" to "Schlösserknacken" due to the german client.

(max(1d20 - (@actor.system.characteristics.in.value + min(@DC, 0)), 0) + max(1d20 - (@actor.system.characteristics.ff.value + min(@DC, 0)),0) + max(1d20 - (@actor.system.characteristics.ff.value + min(@DC, 0)),0)) - (@skills.Schlösserknacken.system.talentValue.value + max(@DC, 0))

Now this appears for an miss:

![image](https://github.com/Saibot393/LocknKey/assets/116867654/df80c28a-8922-4e59-9aae-33d2aedec536)

and this for succsess

![image](https://github.com/Saibot393/LocknKey/assets/116867654/2ccab1d6-012b-46e8-8532-29f1a2d2e865)

Force to open doesnt work anyway

**Comment by Dicio83** [comment]
Is it possible to open the roll-window like if you roll out of the charakter sheet?

Looks like this for lockpicking:
![image](https://github.com/Saibot393/LocknKey/assets/116867654/ee89be7e-1b0e-45a5-bde8-d71e23fff008)

and this for force 

![image](https://github.com/Saibot393/LocknKey/assets/116867654/858ce03f-e5e1-4199-bbaa-dc99865cf5c1)

**Comment by Saibot393** [comment]
Regarding the lock breaking, does it work if you set this as the "Schlossaufbrechen Wurfformel": `(max(1d20 - (@actor.system.characteristics.ko.value + min(@DC, 0)), 0) + max(1d20 - (@actor.system.characteristics.kk.value + min(@DC, 0)),0) + max(1d20 - (@actor.system.characteristics.kk.value + min(@DC, 0)),0)) - (@skills.Kraftakt.system.talentValue.value + max(@DC, 0))`?
The roll window will not be possible due to the system independent way Lock & Key handles rolls, atleast for now.

**Comment by Dicio83** [comment]
That works
![image](https://github.com/Saibot393/LocknKey/assets/116867654/3e9e0e97-6e20-47c3-9532-f82d8af45c14)

**Comment by Dicio83** [comment]
If it is technically possible, it would be great if you could programme it so that you could, for example, press "Alt+ right click" to pick locks and then open the standard window for DSA to roll for lock picking.

![image](https://github.com/Saibot393/LocknKey/assets/116867654/55bcad9b-59ac-4779-8b4b-185c74c3bc61)

There you can then set the DC and take into account any reductions made by a lock pick (e.g. with a simple lock pick, the 2nd subtest for dexterity (ff) is reduced by 1 according to the rules).

![image](https://github.com/Saibot393/LocknKey/assets/116867654/5c49de33-583c-45f1-90d1-1aa1182d6b0b)

Dann rollt man und abhängig vom Ergebnis knackt man die Türe oder nicht.

**Comment by Saibot393** [comment]
I had a look at the underlying code for rolls in DSA and it seems to be relatively complicated, i am not sure if there is a elegant way to add rolls via this window and react correctly react to the results without virtually recoding the whole window.
Getting the level of compatibility with the DSA system up to this point already took a considerable amount of custom code, i will look into adding the boni fields you described for items but i can make no promises on this.

**Comment by Dicio83** [comment]
As it currently stands, I can only enter a default value for a relief and aggravation in the menu of the module, but not in the running game

**Comment by Saibot393** [comment]
what exactly do you mean by "relief and aggravation"?

**Comment by Dicio83** [comment]
> I had a look at the underlying code for rolls in DSA and it seems to be relatively complicated, i am not sure if there is a elegant way to add rolls via this window and react correctly react to the results without virtually recoding the whole window. Getting the level of compatibility with the DSA system up to this point already took a considerable amount of custom code, i will look into adding the boni fields you described for items but i can make no promises on this.

No stress and no rush. They are just ideas and questions as to whether it is even possible. I'm grateful that someone is taking the trouble to develop such cool modules for DSA :-)

**Comment by Dicio83** [comment]
> what exactly do you mean by "relief and aggravation"?

Sorry, I don't think I have the right English words for it :-)

Will try to explain:

**Comment by Saibot393** [comment]
If you want we can write in german

**Comment by Dicio83** [comment]
Alter :-D Nicht dein ernst :-D

**Comment by Saibot393** [comment]
Soll mir recht sein, wenn du willst kannst du mir auf Discord an saibot5279 schreiben

**Comment by Plushtoast** [comment]
Ah auch verpeilt dass es auf deutsch weitergeht

**Comment by Saibot393** [comment]
Mit v2.5.0 sollten die DSA Würfe nun inklusive des Einstellungs Fensters funktionieren (sollte die Einstellung "Nutze System Würfe" aktiviert ist). Noch einmal vielen Dank für den Code.
